### ISSUE TITLE
Lack of tfd.Categorical support on CPU in XLA in float64

### ISSUE THREAD

**Issue by chrism0dwk** [tensorflow/probability#1127]
Hi Guys,

It feels like the following should work in tf(p)-nightly (20201010):
```python
@tf.function(experimental_compile=True)
def foo():
  p = tf.constant([1, 0, 0, 1], dtype=tf.float64)
  X = tfd.Categorical(logits=tf.math.log(p))
  return X.sample(seed=tf.constant([7, 17], tf.int32))

foo()
```
However, I get an `unsupported op: No registered 'StatelessMultinomial' OpKernel for XLA_GPU_JIT` error message.  Colab example [here](https://colab.research.google.com/drive/188Uuhh2xQis7T8e35bsJC35xqUqcoYqA?usp=sharing).   

This only happens if I set a seed in my `sample()` call, making me wonder if it's an issue with the recent changes to the seeding mechanism.

Cheers,

Chris

PS. FWIW I remember before that an omission had been made in the underlying TF/XLA interface that resulted in an op not being compiled for a given dtype.  I wonder if this is the case here?

**Comment by axch** [comment]
Another workaround is to represent the logits in float32 rather than float64, but of course we should get the float64 version working too.

**Comment by chrism0dwk** [comment]
Thanks!  It's worth saying that in my case, the blockage is that I have a complex JointDistribution(Named) that requires double precision fp in one of the component random variables.  This means that the return value of `log_prob` for all components needs to be 64 bit.  I've faked it for now using a tf.Deterministic extra component that does the casting.

**Comment by brianwa84** [comment]
Thanks for the report, Chris. We can look into a fix.

A workaround I've used for other 64/32 cases in the past is writing out a little explicit "Cast" distribution:

```python
class Cast(tfd.Distribution):
  def __init__(self, dist, dtype):
    self.dist = dist
    self.dtype = dtype
    super().__init__(...)
  # def _batch_shape[_tensor], _event_shape[_tensor] if not scalar, simply delegate to self.dist
  def _sample_n(self, n, seed=None):
    return tf.cast(self.dist.sample(n, seed=seed), self.dtype)
  def _log_prob(self, x):
    return self.dist.log_prob(tf.cast(x, self.dist.dtype))
```

**Comment by brianwa84** [comment]
I sent a fix to TF.

**Comment by brianwa84** [comment]
Should be in tomorrow's tf-nightly. Enjoy!

**Comment by chrism0dwk** [comment]
Amazing, thanks Brian.  Very much appreciated!
